### DISCUSSION TITLE
fix/movement_glitch patch branch

### DISCUSSION THREAD

**Discussion by wizardwhosaysni** [ShiningForceCentral/SF2DISASM#4] [Show and tell]
SF2DISASM patch branch available here : 
[https://github.com/ShiningForceCentral/SF2DISASM/tree/fix/movement_glitch](https://github.com/ShiningForceCentral/SF2DISASM/tree/fix/movement_glitch)

### Intro
The movement glitch is used in battles to reach places which are out of the controlled character's moving boundaries.

### Bug explanation :

Controlling the character relies on two decorrelated loops :

* On main CPU thread, the battle character control loop :
  * Assign "controlled entity" script to character's sprite
  * Sprite moves within boundaries until an action button is pressed
  * Action button pressed : if B then go back to initial position, if A/C then check if target position is already occupied
  * **If target position already occupied, then play refusal SFX and restart character control loop, including the entity script assignment : the "controlled entity" execution starts again from the beginning**. To keep in mind with the following explanations.

* At vertical interruption (once every display frame), the controlled character script :
  * A few initialization commands
  * And then the loop with 3 commands :
    1. Command checking is position triggers map block copies (no importance here)
    2. Command checking player input and initiating movement by determining next immediate destination : the next map block.
    3. Command waiting for the next immediate destination to be reached.

**The issue is that when the battle loop restarts the entity loop, the entity loop doesn't wait for the current destination to be reached : it immediately checks input again, based on previous position, and consequently determining a new destination which should not be reachable.**

The entity script : 
```
ROM:00044E3E eas_ControlledCharacter: ac_entityObstructable ON
ROM:00044E42                  ac_mapUncollidable ON
ROM:00044E46                  ac_entityUncollidable ON
ROM:00044E4A                  ac_setSpeed 32,32
ROM:00044E4E                  ac_accelFactors 0,0
ROM:00044E52                  ac_acceleration -2,ON
ROM:00044E56                  ac_deceleration -2,ON
ROM:00044E5A byte_44E5A:      ac_checkMapBlockCopy   
ROM:00044E5C                  ac_controlCharacter
ROM:00044E5E                  ac_waitDest
ROM:00044E60 word_44E60:      ac_branch              
ROM:00044E62                 dc.w (byte_44E5A-word_44E60) & $FFFF
```

### How to reproduce the glitch :
- Initiate movement, entity script is waiting for destination block to be reached.
- Before that, push A/C to exit main character control loop and check is destination block is occupied.
- Block is occupied, go back to loop, re-assign entity script, start the script from its beginning
- Instead of waiting for destination to be reached, player input is listened again first, ready to set a new destination even though current destination is not reached yet.

### Fix implementation : 
**Simply change the command order in the entity script loop !**
New command order :
1. Wait for destination to be reached
2. Block copy check
3. Input read and destination setting
This way, the wait for the destination is never bypassed, and the glitch can't happen anymore.

Fixed entity script : 
```
ROM:00044E3E eas_ControlledCharacter: ac_entityObstructable ON
ROM:00044E42                  ac_mapUncollidable ON
ROM:00044E46                  ac_entityUncollidable ON
ROM:00044E4A                  ac_setSpeed 32,32
ROM:00044E4E                  ac_accelFactors 0,0
ROM:00044E52                  ac_acceleration -2,ON
ROM:00044E56                  ac_deceleration -2,ON
ROM:00044E5A byte_44E5A:      ac_waitDest  
ROM:00044E5C                  ac_checkMapBlockCopy 
ROM:00044E5E                  ac_controlCharacter
ROM:00044E60 word_44E60:      ac_branch              
ROM:00044E62                 dc.w (byte_44E5A-word_44E60) & $FFFF
```

IPS Patch content :
At ROM address 0x44E5A change "00 40 00 02 00 01" with "00 01 00 40 00 02".

Savestate and IPS patch available here : 
[movement_glitch_savestate_and_patch.zip](https://github.com/ShiningForceCentral/SF2DISASM/files/5691153/movement_glitch_savestate_and_patch.zip)

**Comment by ShiningArthur** [comment]
Thank you very much Wiz for that great fix! I just tested your patch and it works like a charm! I can no longer reproduce the movement glitch after the fix! Well done! You must have made a mistake with the ROM address at the bottom. According to your entity script, it is: 0x44E5A.

**Reply by wizardwhosaysni** [reply]
Wow sorry for the late response, but you're completely right ! Fixing my previous message immediately, thank you Arthur !
